cl-postgres/scram.lisp

```diff
@@ -0,0 +1,296 @@
+  (cond ((eq salt-type :string) (setf server-salt (ironclad:ascii-string-to-byte-array server-salt)) )
+        ((eq salt-type :byte-array) t)
+        ((eq salt-type :base64-string) (setf server-salt (cl-base64:base64-string-to-usb8-array server-salt)))
+        ((eq salt-type :hex) (setf server-salt (ironclad:hex-string-to-byte-array server-salt)))
+        (t (cerror "Please enter valid salt-type" "unknown salt-type in gen-salted-password")))
+  (ironclad:pbkdf2-hash-password
+   (ironclad:ascii-string-to-byte-array (saslprep-normalize password))
+   :salt server-salt
+   :digest digest
+   :iterations iterations))
+
+(defun split-server-response (response)
+  "Takes an array of bytes which are encoded in base64,  It should return a list of three alists of the form:
+ ((\"r\" . \"odaUyoz0GpB5GxXLfe2Y8SVjZEosREsxzxhtXY1jiNebxJlohG8IRD1v\")
+ (\"s\" . \"HV25Sl/1VAUF7k+Ddv42dQ==\") (\"i\" . \"4096\") where \"r\" is the server nonce,
+ \"s\" is a base64 encoded salt and \"i\" is the number of iterations for the hash digest.
+
+ We do not use split-sequence building the cons cell because the equal sign can appear in the nonce or salt itself."
+  (loop :for x :in (split-sequence:split-sequence #\, (cl-postgres-trivial-utf-8:utf-8-bytes-to-string response))
+     :collect (let ((split-on (position #\= x)))
+                (cons (subseq x 0 split-on)
+                      (subseq x (1+ split-on))))))
+
+(defun validate-server-nonce (server-nonce client-nonce)
+  "checks whether the server-nonce begins with the client-nonce. Both need to be normal strings."
+  (if (= 0 (search client-nonce server-nonce))
+      t
+      (error 'protocol-violation
+             :message "Client-nonce not found at beginning of server-nonce")))
+
+(defun parse-scram-server-first-response (response client-nonce &key (response-type :base64-usb8-array))
+  "Takes a server response and a client-nonce. If the server response is not in the form of an array of bytes which are encoded in base64, the response type must be specified as either :base64-string or :utf8-string. The client-nonce should be a normal utf8 string.
+It returns the server-response as a normal string, the server-provided-salt as a normal string, and the server-iterations as an integer"
+  (cond ((eq response-type :base64-usb8-array) nil)
+        ((eq response-type :utf8-string) (setf response (cl-base64:base64-string-to-usb8-array (cl-base64:string-to-base64-string response))))
+        ((eq response-type :base64-string) (setf response (cl-base64:base64-string-to-usb8-array response)))
+        (t (cerror "Invalid response type for parse-scram-server-first-response. Must be one of :base64-usb8-array, :utf8-string or :base64-string"
+                   response-type)))
```

Comment:
`cond` of `eq`s with a `cerror` at the end is replaceable with `ccase`.

Reply:
I think ccase prevents control of the error message. Right?

Reply:
Correct. CLHS says:

> The specific error message used by ecase and ccase can vary between implementations. In situations where control of the specific wording of the error message is important, it is better to use case with an otherwise-clause that explicitly signals an error with an appropriate message. 

So, I advise to use `case` instead.

Reply:
Ok.